pkg/reconciler/pipelinerun/resources/workspaces.go

```diff
@@ -0,0 +1,56 @@
+package resources
+
+import (
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+func ResolveWorkspaces(p *v1alpha1.PipelineSpec) (*v1alpha1.PipelineSpec, error) {
+	spec := p.DeepCopy()
+
+	pts := make(map[string]*v1alpha1.PipelineTask)
+	for i := range spec.Tasks {
+		pts[spec.Tasks[i].Name] = &spec.Tasks[i]
+	}
+
+	for taskIdx := range spec.Tasks {
+		pt := &spec.Tasks[taskIdx]
+		for wsIdx := range pt.Workspaces {
+			ws := &pt.Workspaces[wsIdx]
+			if ws.From.Task != "" {
+				resolvedName, err := getResolvedWorkspaceName(pts, ws)
+				if err != nil {
+					return nil, fmt.Errorf("error resolving from clause for workspace %q in pipeline task %q: %w", ws.Name, pt.Name, err)
+				}
+				pt.Workspaces[wsIdx] = v1alpha1.PipelineTaskWorkspaceBinding{
+					Name:      pt.Workspaces[wsIdx].Name,
+					Workspace: resolvedName,
+				}
+			}
+		}
+	}
+	return spec, nil
+}
+
+func getResolvedWorkspaceName(pts map[string]*v1alpha1.PipelineTask, ws *v1alpha1.PipelineTaskWorkspaceBinding) (string, error) {
+SEARCH_FOR_WORKSPACE:
+	for i := 0; i < len(pts); i++ {
+		if ws.Workspace != "" {
+			return ws.Workspace, nil
+		} else if ws.From.Task != "" {
+			if nextTask, ok := pts[ws.From.Task]; ok {
+				for wsi := range nextTask.Workspaces {
+					if nextTask.Workspaces[wsi].Name == ws.From.Name {
+						ws = &nextTask.Workspaces[wsi]
+						continue SEARCH_FOR_WORKSPACE
```

Comment:
im really not sure why Go allows gotos, but imo we should avoid them - maybe we can find a way to make this work with a function call instead?